userspace/libscap/linux/barrier.h

```diff
@@ -17,4 +17,61 @@ limitations under the License.
 */
 #pragma once
 
-#define mem_barrier() __sync_synchronize()
+// This is taken from kernel headers `/include/linux/compiler.h`
+// Used by libpman and scap_bpf engine
+
+#define READ_ONCE(x) (*(volatile typeof(x) *)&x)
+#define WRITE_ONCE(x, v) (*(volatile typeof(x) *)&x) = (v)
+
+#define barrier() asm volatile("" ::: "memory")
+
+#if defined(__x86_64__)
+
+#define smp_mb() asm volatile("lock; addl $0,-132(%%rsp)" ::: "memory", "cc")
+
+#define smp_store_release(p, v) \
+	do {                        \
+		barrier();              \
+		WRITE_ONCE(*p, v);      \
+	} while(0)
+
+#define smp_load_acquire(p)              \
+	({                                   \
+		typeof(*p) ___p = READ_ONCE(*p); \
+		barrier();                       \
+		___p;                            \
+	})
+
+#elif defined(__aarch64__)
+
+#define smp_mb() asm volatile("dmb ish" ::: "memory")
+
+#endif
+
+#ifndef smp_mb
+#define smp_mb() __sync_synchronize()
+#endif
+
+#ifndef smp_store_release
+#define smp_store_release(p, v) \
+	do {                        \
+		smp_mb();               \
+		WRITE_ONCE(*p, v);      \
+	} while(0)
+#endif
+
+#ifndef smp_load_acquire
+#define smp_load_acquire(p)              \
+	({                                   \
+		typeof(*p) ___p = READ_ONCE(*p); \
+		smp_mb();                        \
+		___p;                            \
+	})
+#endif
+
+// This is defined by us
+#if defined(__x86_64__)
```

Comment:
This is used by the scap_kmod and scap_bpf and it uses the same primitive of `smp_store`/`smp_load`